scheds/rust/scx_layered/src/bpf/main.bpf.c

```diff
@@ -1208,6 +1211,79 @@ static bool keep_running(struct cpu_ctx *cctx, struct task_struct *p)
 	return false;
 }
 
+bool timer_antistall(s32 cpu, struct cpu_ctx *cctx){
+	struct task_struct *p;
+	struct task_ctx *tctx;
+	u64 dsq_id, now;
+	u32 idx;
+	
+	if (!cctx || !cctx->layer_idx)
+		return false;
+
+	now = bpf_ktime_get_ns();
+	
+	if(!dsq_timer_check_flag)
+		return false;
+
+	bpf_for(idx, 0, nr_layers) {
+		dsq_id = rotate_layer_id(cctx->layer_idx, idx);
+
+		// check non-fallback dsqs.
+		bpf_for_each(scx_dsq, p, dsq_id, 0) {
+
+			if ((tctx = lookup_task_ctx(p))) {
+				if(tctx->running_at) {
+					if(now - tctx->running_at > NSEC_PER_MSEC * 5) {
+						if(scx_bpf_consume(dsq_id))
+							return true;
+					} else {
+						// if the first entry in a dsq is ok, the dsq is ok.
+						break;
+					}
+				}
+			}
+		}
+	}
+
+	// check fallback dsqs
+	dsq_id = cpu_hi_fallback_dsq_id(cpu);
+	bpf_for_each(scx_dsq, p, dsq_id, 0) {
+		now = bpf_ktime_get_ns();
+
+		if ((tctx = lookup_task_ctx(p))) {
+			if(tctx->running_at) {
+				if(now - tctx->running_at > NSEC_PER_MSEC * 5) {
+					if(scx_bpf_consume(dsq_id))
+						return true;
+				} else {
+					// if the first entry in a dsq is ok, the dsq is ok.
+					break;
+				}
+			}
+		}
+	}
+
+	bpf_for_each(scx_dsq, p, LO_FALLBACK_DSQ, 0) {
+		now = bpf_ktime_get_ns();
+
+		if ((tctx = lookup_task_ctx(p))) {
+			if(tctx->running_at) {
+				if(now - tctx->running_at > NSEC_PER_MSEC * 5) {
+					if(scx_bpf_consume(LO_FALLBACK_DSQ))
+							return true;
+				} else {
+					// if the first entry in a dsq is ok, the dsq is ok.
+					break;
+				}
+			}
+		}
+	}
+
+	// all dsqs are ok, clear check flag.
+	__sync_fetch_and_and(&dsq_timer_check_flag, 0);
```

Comment:
Control reaching here doesn't indicate that there are no stalled tasks. e.g. if a task which is affined to one CPU is stalled and a different CPU ended up running antistall before it, that CPU could reach here and clear the flag before the one which can run the stalled task can run antistall.

The code needs to track whether it saw any stalled task or not separately from consumption succeeding or not.

Reply:
ooh that makes sense wrt/ the outstanding stalls I think, thanks!